[libc] implemented add_function to yaml hdrgen

aaryanshukla opened the pull request:
users can now input a new function in the CLI: the yaml, and .h will
be updated instantly
tested on different cases and does not change the yaml format in anyway
except decreasing extra spaces if present

aaryanshukla committed:
[libc] implemented add_function to yaml hdrgen

users can now input a new function in the CLI: the yaml, and .h will
be updated instantly
tested on different cases and does not change the yaml format in anyway
except decreasing extra spaces if present

alanzhao1 commented on libc/newhdrgen/yaml_to_classes.py:
```diff
@@ -23,6 +22,11 @@
 from class_implementation.classes.object import Object
 
 
+class MyDumper(yaml.Dumper):
```
1. Consider renaming this to something more descriptive (maybe `IndentListDumper`?)
2. If you don't need this class anywhere else, you can define it in `add_function_to_yaml(...)` to reduce its scope.

aaryanshukla committed:
addressed comments regarding naming/scope

michaelrj-google commented on libc/newhdrgen/yaml_to_classes.py:
```diff
@@ -103,16 +107,63 @@ def fill_public_api(header_str, h_def_content):
 
 
-def main(yaml_file, h_def_file, output_dir):
+def add_function_to_yaml(yaml_file, function_details):
```
why is `function_details` not a function object?

aaryanshukla replied:
It was mainly a question of usability. CLI arguments are inherently string-based, so it would make sense to make the CLI a list of strings rather than some type of complex object.

michaelrj-google replied:
I agree that it makes more sense for the user to pass a list of strings, but that list of strings should be used to initialize the `function_details` object.

aaryanshukla committed:
changed class name and moved class

michaelrj-google commented on libc/newhdrgen/yaml_to_classes.py:
```diff
@@ -143,6 +196,19 @@ def main(yaml_file, h_def_file, output_dir):
+        "--add_function",
+        nargs=6,
+        metavar=(
+            "NAME",
```
this should probably be in the same order that you'd find them in a function definition (i.e. `RETURN_TYPE NAME (ARGUMENTS)...`)

aaryanshukla replied:
fixed!

michaelrj-google replied:
this seems like the same order as before, maybe the formatter reordered them?

aaryanshukla committed:
adding a function now uses an object for details

created extra parser function for object to hold values for CL

aaryanshukla committed:
black formatting added

aaryanshukla committed:
fixed ordering of add_function input

RoseZhang03 commented on libc/newhdrgen/yaml_to_classes.py:
```diff
@@ -103,16 +102,92 @@ def fill_public_api(header_str, h_def_content):
+    Returns:
+        Function: An instance of Function initialized with the details.
+    """
+    return_type, name, arguments, standards, guard, attributes = details
```
The order of parameters here is different than the Function class, however this makes more sense. I change the Function class ordering in my next patch.

aaryanshukla commented:
I tested with the function class we have right now and the results are the
same, I don’t think we need to edit the class because we are appending
values so order doesn’t make a difference.

On Mon, Jul 1, 2024 at 10:26 AM RoseZhang03 ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In libc/newhdrgen/yaml_to_classes.py
> <https://github.com/llvm/llvm-project/pull/97079#discussion_r1661346323>:
>
> > +    """
> +    Parse function details from a list of strings and return a Function object.
> +
> +    Args:
> +        details: A list containing function details
> +
> +    Returns:
> +        Function: An instance of Function initialized with the details.
> +    """
> +    return_type, name, arguments, standards, guard, attributes = details
> +    standards = standards.split(",") if standards != "null" else []
> +    arguments = [arg.strip() for arg in arguments.split(",")]
> +    attributes = attributes.split(",") if attributes != "null" else []
> +
> +    return Function(
> +        name=name,
>
> Ordering here is different
> ------------------------------
>
> In libc/newhdrgen/yaml_to_classes.py
> <https://github.com/llvm/llvm-project/pull/97079#discussion_r1661347279>:
>
> > +        name=name,
> +        standards=standards,
> +        return_type=return_type,
> +        arguments=arguments,
> +        guard=guard if guard != "null" else None,
> +        attributes=attributes if attributes else None,
> +    )
> +
> +
> +def add_function_to_yaml(yaml_file, function_details):
> +    """
> +    Add a function to the YAML file.
> +
> +    Args:
> +        yaml_file: The path to the YAML file.
> +        function_details: A list containing function details (return_type, name, arguments, standards, guard, attributes).
>
> Keep consistent ordering. For the patch I will go with this one.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/97079#pullrequestreview-2151904437>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMZZRVDJZ6MDUKHK7PARFMTZKGGL5AVCNFSM6AAAAABKCFYHH6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCNJRHEYDINBTG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

aaryanshukla committed:
changed function ordering to meet function.py

aaryanshukla committed:
fixed formatting issues

michaelrj-google reviewed (approved):
LGTM
